### ISSUE TITLE
Why is quantized graph inference takes much more time than using the original graph?

### ISSUE THREAD

**Issue by yossibiton** [tensorflow/tensorflow#4434]
I followed this [tutorial](https://www.tensorflow.org/versions/r0.10/how_tos/quantization/index.html) in order to quantize my graph into 8 bit.I can't share the exact graph here but i can say it's a simple convolutional neural network.

When i run the [benchmark tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/benchmark) over the original and quantized networks it's clear that the quantized network is much much slower (100 ms vs. 4.5 ms).

Slowest nodes in original network :

```
time average [ms]   [%] [cdf%]  [Op]    [Name]
1.198   26.54%  26.54%  MatMul  fc10/fc10/MatMul
0.337   7.47%   34.02%  Conv2D  conv2/Conv2D
0.332   7.36%   41.37%  Conv2D  conv4/Conv2D
0.323   7.15%   48.53%  Conv2D  conv3/Conv2D
0.322   7.14%   55.66%  Conv2D  conv5/Conv2D
0.310   6.86%   62.53%  Conv2D  conv1/Conv2D
0.118   2.61%   65.13%  Conv2D  conv2_1/Conv2D
0.105   2.32%   67.45%  MaxPool pool1
```

Slowest nodes in quantized network :

```
time average [ms]   [%] [cdf%]  [Op]    [Name]
8.289   47.67%  47.67%  QuantizedMatMul fc10/fc10/MatMul_eightbit_quantized_bias_add
5.398   5.33%   53.00%  QuantizedConv2D conv5/Conv2D_eightbit_quantized_conv
5.248   5.18%   58.18%  QuantizedConv2D conv4/Conv2D_eightbit_quantized_conv
4.981   4.92%   63.10%  QuantizedConv2D conv2/Conv2D_eightbit_quantized_conv
4.908   4.85%   67.95%  QuantizedConv2D conv3/Conv2D_eightbit_quantized_conv
3.167   3.13%   71.07%  QuantizedConv2D conv5_1/Conv2D_eightbit_quantized_conv
3.049   3.01%   74.08%  QuantizedConv2D conv4_1/Conv2D_eightbit_quantized_conv
2.973   2.94%   77.02%  QuantizedMatMul fc11/MatMul_eightbit_quantized_bias_add
```

What is the reason for that ? 
Is it the expected behavior for quantized network ?
### Environment info

Operating System: Ubuntu 16.04
Installed from source, without GPU support :
1. commit hash = 37256f4
2. bazel version = 
Build label: 0.3.1
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Jul 29 09:09:52 2016 (1469783392)
Build timestamp: 1469783392
Build timestamp as int: 1469783392

**Comment by jmchen-g** [comment]
Is it just the first run or consistently slow for many inferences?

**Comment by yossibiton** [comment]
Im using the benchmark tool, which run several inferences.
I also ran it several times and get the same results.

**Comment by yossibiton** [comment]
i will list few simple steps to reproduce the same behavior with inception graph :

1. run my script [classify_image.py](https://gist.github.com/yossibiton/ba571473470824e5fee47c555d01697b) - basically i took [this script](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/imagenet/classify_image.py) and added small changed to support quantized graph and timing. 
It will download the inception graph file and then run forward pass 100 times over an image. I got **average time per forward pass = 785 [ms]**

2. now we will quantize the graph and replace the original graph file (run from you tensorflow repository folder)
`python tensorflow/contrib/quantization/tools/quantize_graph.py --input=/tmp/imagenet/classify_image_graph_def.pb --output_node_names="softmax" --output=/tmp/imagenet/classify_image_graph_q_def.pb --mode=eightbit`

3. we will replace the original graph file with the quantized one :
```
mv /tmp/imagenet/classify_image_graph_def.pb /tmp/imagenet/backup_classify_image_graph_def.pb
mv /tmp/imagenet/classify_image_graph_q_def.pb /tmp/imagenet/classify_image_graph_def.pb 
```

4. running 'classify_image.py' again (now the quantized graph will be used). 
This time i got **average time per forward pass = 6551 [ms]**
That means the quantized graph is about **10x slower**

**Comment by isabel-schwende** [comment]
Actually this has been reported several times before, see here: https://github.com/tensorflow/tensorflow/issues/2807  I've also found the same issue when I played around with 8-bit quantisation. So far it seems that the matrix multiplication as a core operation in the convolution is not optimised yet. It was mentioned in this thread that they are actively working on the problem https://github.com/tensorflow/tensorflow/issues/1592 Since 8-bit quantisation was also mentioned in their new paper on machine translation, I'd expect that the optimised version is being released soon (but this is just an educated guess)

**Comment by jmchen-g** [comment]
@petewarden Could you take a look? Thanks.

**Comment by chenliu0831** [comment]
+1 on this. Any progress? We're also seeing slower performance on CPU.

**Comment by austingg** [comment]
@jmchen-g  @chenliu0831  have your benchmark done on mobile devices ?

**Comment by chenliu0831** [comment]
@austingg Our iOS developer didn't end up benchmarking it given those issues. Instead we are looking into metal to reconstruct the network with trained weight

**Comment by austingg** [comment]
@chenliu0831 great. Metal use gpu, but support only limited operator.

**Comment by mddrill** [comment]
I'm running into this issue also, not only is the network 4 times slower, but it doesn't actually work anymore. I haven't measured the accuracy, but it's a pose estimation algorithm, so I don't need to measure the accuracy to see that it's no longer estimating the pose.

**Comment by clumsydzd** [comment]
@mddrill have you solved your problem? I ran into the same situation

**Comment by mddrill** [comment]
@clumsydzd no I was never able to figure it out.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by oceliktutan** [comment]
Hello, I have a similar problem too. I cannot observe any difference between my optimised graph and my quantised graph in terms of run time. Were you able to solve your problem? Thanks.

**Comment by mddrill** [comment]
Nope. Was never able to do it.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by yifeif** [comment]
@petewarden @suharshs any comment? Thanks!

**Comment by suharshs** [comment]
From https://github.com/tensorflow/tensorflow/issues/2807:

We are focusing our eight-bit efforts on TF Lite (visible at tensorflow/contrib/lite), so we aren't expecting TensorFlow's quantized performance to improve in cases where it's not currently fast. These tend to be on x86 platforms (we're concentrating on ARM performance for mobile), and for models that use ops that we don't have quantized implementations for (which is most models outside a few vision-related ones we've optimized for).

Since we're not likely to see changes in this area soon, I'm closing this as infeasible. Pull requests or other help in this area would be very welcome of course!

**Comment by csuestc** [comment]
@isabel-schwende -- can you please tell me which machine translation paper. I want to see. thanks

**Comment by isabel-schwende** [comment]
There's a section in this paper from 2016 on how they used 8-bit quantization for machine translation https://arxiv.org/abs/1609.08144

**Comment by v1incent** [comment]
Seems that the quantized model is faster on machine with AVX/AVX2 CPU, and tensorflow was built to support AVX/AVX2.
